http-clients/src/main/java/org/triplea/http/client/web/socket/WebSocketConnection.java

```diff
@@ -115,7 +121,15 @@ void connect() {
    */
   void sendMessage(final String message) {
     Preconditions.checkState(!closed);
-    Preconditions.checkState(client.isOpen());
+    if (!client.isOpen()) {
+      try {
+        Awaitility.await()
+            .atMost(connectTimeoutMillis, TimeUnit.MILLISECONDS)
+            .until(() -> client.isOpen());
+      } catch (final ConditionTimeoutException ignored) {
+        throw new CouldNotConnect(serverUri);
+      }
+    }
```

Comment:
IIRC the reason why we introduced the complex logic in the first place was to avoid any kind of blocking behaviour.
Does UI call this method directly? If so I'd actually recommend making `awaitUntilOpen` a separate method, so that the UI can spawn a Thread that waits until the connection is actually open and then enables the UI components to send stuff, whereas this method just gets a precondition that checks if the method is actually open.
This way we could avoid a UI freeze. If this method is invoked on any other Thread besides the EDT, then this is perfectly fine.

Reply:
If UI spawns a connection and then itself issues a 'send', we'll need to block on the 'send'. I don't think we can guarantee that we'll never have that condition. We could try to do something like have 'connect' return a lock token and then require the token to send. IMO just blocking on send is simpler.

Reply:
Some quick research, we do both the 'connect' and 'send' from the EDT. Below is where we call connect (it is through the constructor of the lobby chat transmitter):

```
  public LobbyFrame(final LobbyClient lobbyClient, final ServerProperties serverProperties) {
    super("TripleA Lobby");
    setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
    setIconImage(JFrameBuilder.getGameIcon());
    this.lobbyClient = lobbyClient;
    setJMenuBar(new LobbyMenu(this));
    final ChatTransmitter chatTransmitter = new LobbyChatTransmitter(lobbyClient);
    final Chat chat = new Chat(chatTransmitter);
```
 
Then the 'connect' message below is the first time we send data over websocket:
```
 public Chat(final ChatTransmitter chatTransmitter) {
    this.localPlayerName = chatTransmitter.getLocalPlayerName();
    this.chatTransmitter = chatTransmitter;
    chatTransmitter.setChatClient(this);
    sentMessagesHistory = new SentMessagesHistory();
    Optional.ofNullable(chatTransmitter.connect()).ifPresent(chatters::addAll);
    updateConnections();
  }
```